아이디, 패스워드 로그인 구현

minju412 opened the pull request:
## 📄 Summary
#40 

<br>

## 🙋‍♀️ To Reviewers
- TokenProvider는 따로 클래스로 구현하지 않고 AuthenticationFilter 클래스 내부 로직으로 포함되어 있습니다!
- 아마 충돌이 많이 날 것 같은데 같이 해결해욥!

minju412 committed:
feat: 회원가입 api 추가

- 이메일, 비밀번호, 닉네임을 이용해 회원가입
- 비밀번호는 암호화하여 데이터베이스에 저장
- 회원가입 시 Member 테이블에 기본 정보가 저장되며 BasicLoginMember 테이블에 암호화된 비밀번호 저장
  - 소셜로그인 회원은 비밀번호가 필요 없기 때문에 테이블 분리

minju412 committed:
feat: 로그인 기능 구현

- 로그인 시 Authorization 헤더로 액세스 토큰 값을 넘겨주고, 쿠키로 리프레시 토큰 값을 전달합니다.
- SecurityConfig에서 권한 설정을 통해 특정 권한을 가진 회원만 접근할 수 있도록 설정합니다.
- JWT Secret 등 민감한 설정 정보는 JWTProperties라는 상수 클래스를 만들었고 gitignore 처리하였습니다.

chwangmin commented on src/main/java/com/ssafy/naem/global/jwt/filter/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,92 @@
+		// Jwt 토큰 복호화
+		Claims claims = parseClaims(accessToken);
+
+		if (claims.get("auth") == null) {
+			throw new RuntimeException("권한 정보가 없는 토큰입니다.");
+		}
```
여기서 controllerAdvice에서 RuntimeException을 잡지 않아서 원하시는 response가 만들어지지 않을 거 같아요!
왜냐하면 현재 filter 구조가 servelt 전에서 작동하기 때문이라고 생각해요
servletResponse 세팅을 적절히 다시 만들어보는 건 어떨까요?
[참고](https://jhkimmm.tistory.com/29)

minju412 replied:
몰랐던 사실이네요..!😲 참고 링크까지 감사합니닷,, 공부해서 다음 pr에 반영할게요!

chwangmin commented on src/main/java/com/ssafy/naem/global/jwt/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,106 @@
+			.map(GrantedAuthority::getAuthority)
+			.collect(Collectors.joining(","));
+
+		// Access Token 생성
+		String accessToken = Jwts.builder()
+			.setSubject(authResult.getName()) // email
+			.claim("auth", authorities)
```
jwt token 객체를 따로 생성해서 구현하는 건 어떨까요?
왜냐하면 accessToken을 생성하는 코드를 filter에서 사용하지 않을 경우가 있다면, 중복되는 코드가 발생할 것 같아서요

minju412 replied:
좋습니다! dto 객체를 따로 구현하겠습니닷

minju412 committed:
fix: 액세스 토큰 생성 시 무조건 GENERAL 권한을 넣는 코드 제거

- CustomUserDetailsService의 createUserDetails 메서드 내부에서 authorities에 일반 회원 권한을 넣는 코드를 삭제합니다.
- 권한이 없는 회원의 경우 예외를 발생시킵니다.

chwangmin reviewed (approved):
security filter로 인가 인증 구현하시느라 고생하셨습니다.

minju412 committed:
refactor: JwtTokenResponse 객체 활용

- JwtTokenResponse라는 DTO 객체를 통해 전달하고자 하는 값을 감싸서 전달합니다.

bjho606 commented:
인증, 인가의 처리를 Interceptor 에서 하는 것과 Spring Security에서 하는 것의 차이와 이유에 대해서 생각해보면 좋을것 같습니다. (저는 아직 잘 몰라서요..ㅎㅎ)
[참고](https://velog.io/@slolee/Spring-Security-%EA%B7%B8%EB%A0%87%EA%B2%8C-%EC%93%B0%EC%A7%80-%EB%A7%88%EC%84%B8%EC%9A%94)

minju412 committed:
refactor: JwtTokenProvider 클래스 추가

- 액세스 토큰과 리프레시 토큰 관련 유틸 클래스인 JwtTokenProvider를 추가했습니다.
- 기존에는 JwtAuthenticationFilter 내에서 토큰 생성 로직을 했다면, 이를 분리함으로써 코드 재사용성을 높일 수 있습니다.

minju412 committed:
Merge branch 'develop' into feature/#40-basic-login
